presto-main/src/main/java/io/prestosql/operator/scalar/time/TimeOperators.java

```diff
@@ -0,0 +1,298 @@
+            @LiteralParameter("sourcePrecision") long sourcePrecision,
+            @LiteralParameter("targetPrecision") long targetPrecision,
+            @SqlType("time(sourcePrecision)") long time)
+    {
+        if (sourcePrecision <= targetPrecision) {
+            return time;
+        }
+
+        return round(time, (int) (MAX_PRECISION - targetPrecision)) % PICOSECONDS_PER_DAY;
+    }
+
+    @ScalarOperator(ADD)
+    @LiteralParameters({"p", "u"})
+    @SqlType("time(u)")
+    @Constraint(variable = "u", expression = "max(3, p)") // interval is currently p = 3
+    public static long timePlusIntervalDayToSecond(@SqlType("time(p)") long time, @SqlType(StandardTypes.INTERVAL_DAY_TO_SECOND) long interval)
+    {
+        return add(time, interval * PICOSECONDS_PER_MILLISECOND);
+    }
+
+    @ScalarOperator(ADD)
+    @LiteralParameters({"p", "u"})
+    @SqlType("time(u)")
+    @Constraint(variable = "u", expression = "max(3, p)") // interval is currently p = 3
+    public static long intervalDayToSecondPlusTime(@SqlType(StandardTypes.INTERVAL_DAY_TO_SECOND) long interval, @SqlType("time(p)") long time)
+    {
+        return timePlusIntervalDayToSecond(time, interval);
+    }
+
+    @ScalarOperator(SUBTRACT)
+    @LiteralParameters({"p", "u"})
+    @SqlType("time(u)")
+    @Constraint(variable = "u", expression = "max(3, p)") // interval is currently p = 3
+    public static long timeMinusIntervalDayToSecond(@SqlType("time(p)") long time, @SqlType(StandardTypes.INTERVAL_DAY_TO_SECOND) long interval)
+    {
+        return add(time, -interval * PICOSECONDS_PER_MILLISECOND);
+    }
+
+    @ScalarOperator(CAST)
+    @LiteralParameters({"x", "p"})
+    @SqlType("varchar(x)")
+    public static Slice castToVarchar(@LiteralParameter("p") long precision, @SqlType("time(p)") long value)
+    {
+        int size = (int) (8 + // hour:minute:second
+                        (precision > 0 ? 1 : 0) + // period
+                        precision); // fraction
+
+        DynamicSliceOutput output = new DynamicSliceOutput(size);
+
+        output.appendBytes(format("%02d", value / PICOSECONDS_PER_HOUR).getBytes(UTF_8));
+        output.appendByte(':');
+        output.appendBytes(format("%02d", (value / PICOSECONDS_PER_MINUTE) % MINUTES_PER_HOUR).getBytes(UTF_8));
+        output.appendByte(':');
+        output.appendBytes(format("%02d", (value / PICOSECONDS_PER_SECOND) % SECONDS_PER_MINUTE).getBytes(UTF_8));
+
+        if (precision > 0) {
+            long scaledFraction = (value % PICOSECONDS_PER_SECOND) / scaleFactor((int) precision, MAX_PRECISION);
+            output.appendByte('.');
+            output.appendBytes(format("%0" + precision + "d", scaledFraction).getBytes(UTF_8));
+        }
+
+        return output.slice();
+    }
+
+    public static long add(long time, long delta)
+    {
+        long result = (time + delta) % PICOSECONDS_PER_DAY;
+        if (result < 0) {
+            result += PICOSECONDS_PER_DAY;
```

Comment:
As commented earlier, the `if` here may be insufficient for large negative delta

Reply:
See ` % PICOSECONDS_PER_DAY` in the line above.